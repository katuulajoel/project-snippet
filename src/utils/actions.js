/**
 * This get called before an async function is executed
 * @param {string} type
 * @returns
 */
export function start(type) {
  return { type };
}

/**
 * This dispatches successful Promises
 * @param {string} type
 * @param {*} data
 * @returns
 */
export function success(type, data) {
  return { type, data };
}

/**
 * For dispatching failed Promises
 * @param {string} type
 * @param {object} param1
 * @returns
 */
export function failed(type, { message }) {
  const error = message;
  return { type, error };
}
