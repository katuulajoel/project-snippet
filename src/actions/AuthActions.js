import axios from "axios";
import {
  ENDPOINT_LOGIN,
  ENDPOINT_LOGOUT,
  ENDPOINT_VERIFY,
  ENDPOINT_REGISTER,
  ENDPOINT_APPLY,
  ENDPOINT_RESET_PASSWORD,
  ENDPOINT_RESET_PASSWORD_CONFIRM,
  ENDPOINT_EMAIL_VISITOR,
} from "./utils/api";

import {
  sendGAEvent,
  sendTwitterSignUpEvent,
  GA_EVENT_CATEGORIES,
  GA_EVENT_ACTIONS,
  GA_EVENT_LABELS,
  AUTH_METHODS,
  getGAUserType,
  getUserTypeTwitter,
} from "./utils/tracking";
import { getUser } from "../components/utils/auth";

export const LOGIN_START = "LOGIN_START";
export const LOGIN_SUCCESS = "LOGIN_SUCCESS";
export const LOGIN_FAILED = "LOGIN_FAILED";
export const VERIFY_START = "VERIFY_START";
export const VERIFY_SUCCESS = "VERIFY_SUCCESS";
export const VERIFY_FAILED = "VERIFY_FAILED";
export const LOGOUT_START = "LOGOUT_START";
export const LOGOUT_SUCCESS = "LOGOUT_SUCCESS";
export const LOGOUT_FAILED = "LOGOUT_FAILED";
export const REGISTER_START = "REGISTER_START";
export const REGISTER_SUCCESS = "REGISTER_SUCCESS";
export const REGISTER_FAILED = "REGISTER_FAILED";
export const APPLY_START = "APPLY_START";
export const APPLY_SUCCESS = "APPLY_SUCCESS";
export const APPLY_FAILED = "APPLY_FAILED";
export const RETRIEVE_APPLICATION_START = "RETRIEVE_APPLICATION_START";
export const RETRIEVE_APPLICATION_SUCCESS = "RETRIEVE_APPLICATION_SUCCESS";
export const RETRIEVE_APPLICATION_FAILED = "RETRIEVE_APPLICATION_FAILED";
export const RESET_PASSWORD_START = "RESET_PASSWORD_START";
export const RESET_PASSWORD_SUCCESS = "RESET_PASSWORD_SUCCESS";
export const RESET_PASSWORD_FAILED = "RESET_PASSWORD_FAILED";
export const RESET_PASSWORD_CONFIRM_START = "RESET_PASSWORD_CONFIRM_START";
export const RESET_PASSWORD_CONFIRM_SUCCESS = "RESET_PASSWORD_CONFIRM_SUCCESS";
export const RESET_PASSWORD_CONFIRM_FAILED = "RESET_PASSWORD_CONFIRM_FAILED";
export const AUTH_REDIRECT = "AUTH_REDIRECT";
export const EMAIL_VISITOR_AUTH_START = "EMAIL_VISITOR_AUTH_START";
export const EMAIL_VISITOR_AUTH_SUCCESS = "EMAIL_VISITOR_AUTH_SUCCESS";
export const EMAIL_VISITOR_AUTH_FAILED = "EMAIL_VISITOR_AUTH_FAILED";
export const EMAIL_VISITOR_VERIFY_START = "EMAIL_VISITOR_VERIFY_START";
export const EMAIL_VISITOR_VERIFY_SUCCESS = "EMAIL_VISITOR_VERIFY_SUCCESS";
export const EMAIL_VISITOR_VERIFY_FAILED = "EMAIL_VISITOR_VERIFY_FAILED";

const arr = [
  "Good to see you back!",
  "How are you doing today?",
  "Great to have you here!",
  "What’s cooking, chef?",
  "We’ve missed you!",
  "Have a great day!",
  "Hello, world!",
  "Press Any Key to Continue…",
  "Have you tried restarting?",
  "You light my Fire(wall)",
];

export function authenticate(credentials) {
  return (dispatch) => {
    dispatch(authStart(credentials));
    return axios
      .post(ENDPOINT_LOGIN, credentials)
      .then(function (response) {
        dispatch(authSuccess(response.data));
        return response.data;
      })
      .catch(function (error) {
        dispatch(authFailed(error.response ? error.response.data : null));
      });
  };
}

export function authStart(credentials) {
  return {
    type: LOGIN_START,
    credentials,
  };
}

export function authSuccess(data) {
  let index = Math.floor(Math.random() * arr.length);
  let user = Object.assign({}, data.user, { welcomeMessage: arr[index] });
  window.localStorage.setItem("welcomeMsg", arr[index]);

  sendGAEvent(
    GA_EVENT_CATEGORIES.AUTH,
    GA_EVENT_ACTIONS.SIGN_IN,
    getGAUserType(user)
  );
  return {
    type: LOGIN_SUCCESS,
    user,
  };
}

export function authFailed(error) {
  return {
    type: LOGIN_FAILED,
    error,
  };
}

export function authenticateEmailVisitor(credentials) {
  return (dispatch) => {
    dispatch(authEmailVisitorStart(credentials));
    axios
      .post(ENDPOINT_EMAIL_VISITOR, credentials)
      .then(function (response) {
        dispatch(authEmailVisitorSuccess(response.data));
      })
      .catch(function (error) {
        dispatch(
          authEmailVisitorFailed(error.response ? error.response.data : null)
        );
      });
  };
}

export function authEmailVisitorStart(credentials) {
  return {
    type: EMAIL_VISITOR_AUTH_START,
    credentials,
  };
}

export function authEmailVisitorSuccess(visitor) {
  sendGAEvent(GA_EVENT_CATEGORIES.AUTH, GA_EVENT_ACTIONS.BROWSE_DEVS);
  return {
    type: EMAIL_VISITOR_AUTH_SUCCESS,
    visitor,
  };
}

export function authEmailVisitorFailed(error) {
  return {
    type: EMAIL_VISITOR_AUTH_FAILED,
    error,
  };
}

export function verify() {
  return (dispatch) => {
    dispatch(verifyStart());
    axios
      .get(ENDPOINT_VERIFY)
      .then(function (response) {
        dispatch(verifySuccess(response.data));
      })
      .catch(function () {
        dispatch(verifyEmailVisitor());
      });
  };
}

export function verifyStart() {
  return {
    type: VERIFY_START,
  };
}

export function verifySuccess(data) {
  let user = Object.assign({}, data, {
    welcomeMessage: window.localStorage.getItem("welcomeMsg"),
  });
  return {
    type: VERIFY_SUCCESS,
    user,
  };
}

export function verifyFailed(error) {
  return {
    type: VERIFY_FAILED,
    error,
  };
}

export function verifyEmailVisitor() {
  return (dispatch) => {
    dispatch(verifyEmailVisitorStart());
    axios
      .get(ENDPOINT_EMAIL_VISITOR)
      .then(function (response) {
        dispatch(verifyEmailVisitorSuccess(response.data));
      })
      .catch(function (error) {
        dispatch(
          verifyEmailVisitorFailed(error.response ? error.response.data : null)
        );
      });
  };
}

export function verifyEmailVisitorStart() {
  return {
    type: EMAIL_VISITOR_VERIFY_START,
  };
}

export function verifyEmailVisitorSuccess(visitor) {
  return {
    type: EMAIL_VISITOR_VERIFY_SUCCESS,
    visitor,
  };
}

export function verifyEmailVisitorFailed(error) {
  return {
    type: EMAIL_VISITOR_VERIFY_FAILED,
    error,
  };
}

export function logout() {
  return (dispatch) => {
    dispatch(logoutStart());
    axios
      .post(ENDPOINT_LOGOUT, {})
      .then(function () {
        dispatch(logoutSuccess());
      })
      .catch(function (error) {
        dispatch(logoutFailed(error.response ? error.response.data : null));
      });
  };
}

export function logoutStart() {
  return {
    type: LOGOUT_START,
  };
}

export function logoutSuccess() {
  window.localStorage.removeItem("welcomeMsg");
  sendGAEvent(
    GA_EVENT_CATEGORIES.AUTH,
    GA_EVENT_ACTIONS.LOG_OUT,
    getGAUserType(getUser())
  );
  return {
    type: LOGOUT_SUCCESS,
  };
}

export function logoutFailed(error) {
  return {
    type: LOGOUT_FAILED,
    error,
  };
}

export function register(details) {
  return (dispatch) => {
    dispatch(registerStart(details));
    axios
      .post(ENDPOINT_REGISTER, details)
      .then(function (response) {
        dispatch(registerSuccess(response.data));

        let user_type = getUserTypeTwitter(details.type),
          method = AUTH_METHODS.EMAIL;
        sendTwitterSignUpEvent({ user_type, method });
      })
      .catch(function (error) {
        dispatch(registerFailed(error.response ? error.response.data : null));
      });
  };
}

export function registerStart(details) {
  return {
    type: REGISTER_START,
    details,
  };
}

export function registerSuccess(data) {
  let user = data.user;
  sendGAEvent(
    GA_EVENT_CATEGORIES.REGISTRATION,
    GA_EVENT_ACTIONS.SIGN_UP,
    getGAUserType(user)
  );
  return {
    type: REGISTER_SUCCESS,
    user,
  };
}

export function registerFailed(error) {
  return {
    type: REGISTER_FAILED,
    error,
  };
}

export function apply(details) {
  return (dispatch) => {
    dispatch(applyStart(details));
    axios
      .post(ENDPOINT_APPLY, details)
      .then(function (response) {
        dispatch(applySuccess(response.data));
      })
      .catch(function (error) {
        dispatch(applyFailed(error.response ? error.response.data : null));
      });
  };
}

export function applyStart(details) {
  return {
    type: APPLY_START,
    details,
  };
}

export function applySuccess(application) {
  sendGAEvent(
    GA_EVENT_CATEGORIES.AUTH,
    GA_EVENT_ACTIONS.DEV_APPLY,
    GA_EVENT_LABELS.DEVELOPER
  );
  return {
    type: APPLY_SUCCESS,
    application,
  };
}

export function applyFailed(error) {
  return {
    type: APPLY_FAILED,
    error,
  };
}

export function retrieveApplication(key) {
  return (dispatch) => {
    dispatch(retrieveApplicationStart(key));
    axios
      .get(ENDPOINT_APPLY + "key/" + key + "/")
      .then(function (response) {
        dispatch(retrieveApplicationSuccess(response.data));
      })
      .catch(function (error) {
        dispatch(
          retrieveApplicationFailed(error.response ? error.response.data : null)
        );
      });
  };
}

export function retrieveApplicationStart(key) {
  return {
    type: RETRIEVE_APPLICATION_START,
    key,
  };
}

export function retrieveApplicationSuccess(application) {
  return {
    type: RETRIEVE_APPLICATION_SUCCESS,
    application,
  };
}

export function retrieveApplicationFailed(error) {
  return {
    type: RETRIEVE_APPLICATION_FAILED,
    error,
  };
}

export function resetPassword(email) {
  return (dispatch) => {
    dispatch(resetPasswordStart(email));
    axios
      .post(ENDPOINT_RESET_PASSWORD, email)
      .then(function (response) {
        dispatch(resetPasswordSuccess(response.data));
      })
      .catch(function (error) {
        dispatch(
          resetPasswordFailed(error.response ? error.response.data : null)
        );
      });
  };
}

export function resetPasswordStart(email) {
  return {
    type: RESET_PASSWORD_START,
    email,
  };
}

export function resetPasswordSuccess(response) {
  sendGAEvent(GA_EVENT_CATEGORIES.AUTH, GA_EVENT_ACTIONS.RECOVER_PASSWORD);
  return {
    type: RESET_PASSWORD_SUCCESS,
    response,
  };
}

export function resetPasswordFailed(error) {
  return {
    type: RESET_PASSWORD_FAILED,
    error,
  };
}

export function resetPasswordConfirm(credentials) {
  return (dispatch) => {
    dispatch(resetPasswordConfirmStart(credentials));
    axios
      .post(ENDPOINT_RESET_PASSWORD_CONFIRM, credentials)
      .then(function (response) {
        dispatch(resetPasswordConfirmSuccess(response.data));
      })
      .catch(function (error) {
        dispatch(
          resetPasswordConfirmFailed(
            error.response ? error.response.data : null
          )
        );
      });
  };
}

export function resetPasswordConfirmStart(credentials) {
  return {
    type: RESET_PASSWORD_CONFIRM_START,
    credentials,
  };
}

export function resetPasswordConfirmSuccess(response) {
  sendGAEvent(
    GA_EVENT_CATEGORIES.AUTH,
    GA_EVENT_ACTIONS.RECOVER_PASSWORD_CONFIRM
  );
  return {
    type: RESET_PASSWORD_CONFIRM_SUCCESS,
    response,
  };
}

export function resetPasswordConfirmFailed(error) {
  return {
    type: RESET_PASSWORD_CONFIRM_FAILED,
    error,
  };
}

export function authRedirect(path) {
  return {
    type: AUTH_REDIRECT,
    path,
  };
}
