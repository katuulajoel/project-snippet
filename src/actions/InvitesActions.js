import axios from "axios";
import { ENDPOINT_INVITE } from "./utils/api";
import * as actions from "./utils/ActionTypes";

export function getPendingInvites(selection) {
  return function (dispatch) {
    dispatch(getPendingInvitesStart(selection));
    axios
      .get(ENDPOINT_INVITE, { params: { used: "False" } })
      .then(function (response) {
        dispatch(getPendingInvitesSuccess(response.data, selection));
      })
      .catch(function (error) {
        dispatch(
          getPendingInvitesFailed(
            error.response ? error.response.data : null,
            selection
          )
        );
      });
  };
}

export function getPendingInvitesStart(selection) {
  return {
    type: actions.GET_PENDING_INVITES_START,
    selection,
  };
}

export function getPendingInvitesSuccess(data) {
  console.log(data);
  return {
    type: actions.GET_PENDING_INVITES_SUCCESS,
    payload: data,
  };
}

export function getPendingInvitesFailed(error, selection) {
  return {
    type: actions.GET_PENDING_INVITES_FAILED,
    error,
    selection,
  };
}

export function getMorePendingInvites(selection, url) {
  return (dispatch) => {
    dispatch(getMorePendingInvitesStart(url, selection));
    axios
      .get(url)
      .then(function (response) {
        dispatch(getMorePendingInvitesSuccess(response.data, selection));
      })
      .catch(function (error) {
        dispatch(
          getMorePendingInvitesFailed(
            error.response ? error.response.data : null,
            selection
          )
        );
      });
  };
}

export function getMorePendingInvitesStart(url, selection) {
  return {
    type: actions.GET_MORE_PENDING_INVITES_START,
    url,
    selection,
  };
}

export function getMorePendingInvitesSuccess(response, selection) {
  return {
    type: actions.GET_MORE_PENDING_INVITES_SUCCESS,
    items: response.results,
    previous: response.previous,
    next: response.next,
    count: response.count,
    selection,
  };
}

export function getMorePendingInvitesFailed(error) {
  return {
    type: actions.GET_MORE_PENDING_INVITES_FAILED,
    error,
  };
}

export function deleteInvite(id, selectionKey) {
  return (dispatch) => {
    dispatch(deleteInviteStart(id));
    axios
      .delete(`${ENDPOINT_INVITE}${id}/`, {})
      .then(function () {
        dispatch(getPendingInvites(selectionKey));
        dispatch(deleteInviteSuccess(id));
      })
      .catch(function (error) {
        dispatch(
          deleteInviteFailed(error.response ? error.response.data : null)
        );
      });
  };
}

export function invite(data, selectionKey, type = null) {
  return (dispatch) => {
    dispatch(inviteStart(data, selectionKey));
    let request_method = type ? "patch" : "post";
    axios
      .request({
        url: `${ENDPOINT_INVITE}${
          request_method === "patch" ? data.id + "/" : ""
        }`,
        method: request_method,
        data,
      })
      .then(function () {
        // dispatch(inviteSuccess(response.data, selectionKey));
      })
      .catch(function (error) {
        dispatch(
          inviteFailed(
            error.response ? error.response.data : null,
            selectionKey
          )
        );
      });
  };
}

export function inviteStart(details, selectionKey) {
  return {
    type: actions.INVITE_START,
    details,
    selectionKey,
  };
}

// export function inviteSuccess(invite, selectionKey) {
//   sendGAEvent(
//     GA_EVENT_CATEGORIES.AUTH,
//     GA_EVENT_ACTIONS.DEV_INVITE,
//     getGAUserType(getUser())
//   );
//   return {
//     type: actions.INVITE_SUCCESS,
//     invite,
//     selectionKey,
//   };
// }

export function inviteFailed(error, selectionKey) {
  return {
    type: actions.INVITE_FAILED,
    error,
    selectionKey,
  };
}

export function deleteInviteStart(id) {
  return {
    type: actions.DELETE_INVITE_START,
    id,
  };
}

export function deleteInviteSuccess(id) {
  return {
    type: actions.DELETE_INVITE_SUCCESS,
    id,
  };
}

export function deleteInviteFailed(error) {
  return {
    type: actions.DELETE_INVITE_FAILED,
    error,
  };
}
