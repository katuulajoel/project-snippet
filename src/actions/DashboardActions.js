/* eslint-disable no-unused-vars */
import axios from "axios";
import { ENDPOINT_NOTIFICATIONS, ENDPOINT_NOTIFICATION_LOG } from "./utils/api";

import {
  FETCH_NOTIFICATIONS_START,
  FETCH_NOTIFICATIONS_SUCCESS,
  FETCH_NOTIFICATIONS_FAILED,
  DELETE_NOTIFICATIONS_START,
  DELETE_NOTIFICATIONS_SUCCESS,
  DELETE_NOTIFICATIONS_FAILED,
  CREATE_NOTIFICATION_LOG_START,
  CREATE_NOTIFICATION_LOG_SUCCESS,
  CREATE_NOTIFICATION_LOG_FAILED,
} from "./utils/ActionTypes";
import { success, start, failed } from "./utils/actions";

export function getNotifications() {
  return (dispatch) => {
    dispatch(start(FETCH_NOTIFICATIONS_START));
    axios
      .get(ENDPOINT_NOTIFICATIONS)
      .then(function (response) {
        dispatch(success(FETCH_NOTIFICATIONS_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(FETCH_NOTIFICATIONS_FAILED, error));
      });
  };
}

export function createNotificationLog(notification_log) {
  return (dispatch) => {
    dispatch(start(CREATE_NOTIFICATION_LOG_START));
    axios
      .post(ENDPOINT_NOTIFICATION_LOG, notification_log)
      .then(function (response) {
        dispatch(success(CREATE_NOTIFICATION_LOG_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(CREATE_NOTIFICATION_LOG_FAILED, error));
      });
  };
}

export function clearNotification() {
  return (dispatch) => {
    dispatch(start(DELETE_NOTIFICATIONS_START));
    axios
      .delete(ENDPOINT_NOTIFICATIONS)
      .then(function (response) {
        dispatch(success(DELETE_NOTIFICATIONS_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(DELETE_NOTIFICATIONS_FAILED, error));
      });
  };
}
