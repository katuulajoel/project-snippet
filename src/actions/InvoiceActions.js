import axios from "axios";
import { ENDPOINT_INVOICES, ENDPOINT_INVOICES_SUMMARY } from "./utils/api";
import * as types from "./utils/ActionTypes";
import { success, start, failed } from "./utils/actions";

export function createInvoice(invoice) {
  return (dispatch) => {
    dispatch(start(types.CREATE_INVOICE_START));
    axios
      .post(ENDPOINT_INVOICES, invoice)
      .then(function (response) {
        dispatch(success(types.CREATE_INVOICE_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(types.CREATE_INVOICE_FAILED, error));
      });
  };
}

export function createInvoiceBatch(invoices) {
  return (dispatch) => {
    dispatch(start(types.CREATE_INVOICE_BATCH_START));
    axios
      .post(`${ENDPOINT_INVOICES}bulk/`, invoices)
      .then(function (response) {
        dispatch(success(types.CREATE_INVOICE_BATCH_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(types.CREATE_INVOICE_BATCH_FAILED, error));
      });
  };
}

export function downloadInoicesCsv(invoiceSet) {
  return (dispatch) => {
    dispatch(start(types.DOWNLOAD_INVOICE_CSV_STARTED));
    axios
      .get(`${ENDPOINT_INVOICES}export/`, { params: invoiceSet })
      .then(function ({ data }) {
        dispatch(success(types.DOWNLOAD_INVOICE_CSV_SUCCESS, data));
      })
      .catch(function (error) {
        dispatch(failed(types.DOWNLOAD_INVOICE_CSV_FAILED, error));
      });
  };
}

export function listInvoices(filter, search = false) {
  return (dispatch) => {
    dispatch(
      start(search ? types.SEARCH_INVOICES_START : types.LIST_INVOICES_START)
    );
    // TODO: refactor even further
    axios
      .get(
        filter.archived == "True"
          ? ENDPOINT_INVOICES + "archived/"
          : ENDPOINT_INVOICES,
        { params: filter }
      )
      .then(function (response) {
        dispatch(
          success(
            search
              ? types.SEARCH_INVOICES_SUCCESS
              : types.LIST_INVOICES_SUCCESS,
            response.data
          )
        );
      })
      .catch(function (error) {
        dispatch(
          failed(
            search ? types.SEARCH_INVOICES_FAILED : types.LIST_INVOICES_FAILED,
            error
          )
        );
      });
  };
}

export function retrieveInvoice(id) {
  return (dispatch) => {
    dispatch(start(types.RETRIEVE_INVOICE_START));
    axios
      .get(ENDPOINT_INVOICES + id + "/")
      .then(function (response) {
        dispatch(success(types.RETRIEVE_INVOICE_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(types.RETRIEVE_INVOICE_FAILED, error));
      });
  };
}

export function updateInvoice(id, invoice) {
  return (dispatch) => {
    dispatch(start(types.UPDATE_INVOICE_START));
    axios
      .patch(ENDPOINT_INVOICES + id + "/", invoice)
      .then(function (response) {
        dispatch(success(types.UPDATE_INVOICE_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(types.UPDATE_INVOICE_FAILED, error));
      });
  };
}

export function listMoreInvoices(url, search = false) {
  return (dispatch) => {
    dispatch(
      start(
        search
          ? types.SEARCH_MORE_INVOICES_START
          : types.LIST_MORE_INVOICES_START
      )
    );
    axios
      .get(url)
      .then(function (response) {
        dispatch(
          success(
            search
              ? types.SEARCH_MORE_INVOICES_SUCCESS
              : types.LIST_MORE_INVOICES_SUCCESS,
            response.data
          )
        );
      })
      .catch(function (error) {
        dispatch(
          failed(
            search
              ? types.SEARCH_MORE_INVOICES_FAILED
              : types.LIST_MORE_INVOICES_FAILED,
            error
          )
        );
      });
  };
}

export function deleteInvoice(id) {
  return (dispatch) => {
    dispatch(start(types.DELETE_INVOICE_START));
    axios
      .delete(ENDPOINT_INVOICES + id + "/")
      .then(function () {
        dispatch(success(types.DELETE_INVOICE_SUCCESS, { id }));
      })
      .catch(function (error) {
        dispatch(failed(types.DELETE_INVOICE_FAILED, error));
      });
  };
}

export function archiveInvoice(id) {
  return (dispatch) => {
    dispatch(start(types.ARCHIVE_INVOICE_START));
    axios
      .post(`${ENDPOINT_INVOICES}${id}/archive/`)
      .then(function () {
        dispatch(success(types.ARCHIVE_INVOICE_SUCCESS, { id }));
      })
      .catch(function (error) {
        dispatch(failed(types.ARCHIVE_INVOICE_FAILED, error));
      });
  };
}

export function generateInvoice(id) {
  return (dispatch) => {
    dispatch(start(types.GENERATE_INVOICE_START));
    axios
      .post(`${ENDPOINT_INVOICES}${id}/generate/`)
      .then(function (response) {
        dispatch(success(types.GENERATE_INVOICE_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(types.GENERATE_INVOICE_FAILED, error));
      });
  };
}

export function bulkAction(invoices, action) {
  return (dispatch) => {
    dispatch(start(types.BULK_ACTION_START));
    axios
      .patch(`${ENDPOINT_INVOICES}bulk/edit/`, { invoices, action })
      .then(function () {
        dispatch(success(types.BULK_ACTION_SUCCESS, {}));
      })
      .catch(function (error) {
        dispatch(failed(types.BULK_ACTION_FAILED, error));
      });
  };
}

export function getInvoiceSummary(params) {
  return (dispatch) => {
    dispatch(start(types.INVOICE_SUMMARY_START));
    axios
      .get(`${ENDPOINT_INVOICES_SUMMARY}`, { params: params })
      .then(function (response) {
        dispatch(success(types.INVOICE_SUMMARY_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(types.INVOICE_SUMMARY_FAILED, error));
      });
  };
}
