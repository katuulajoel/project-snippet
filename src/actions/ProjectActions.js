/* eslint-disable no-unused-vars */
import axios from "axios";
import { ENDPOINT_PROJECTS } from "./utils/api";

import {
  FETCH_PROJECT_START,
  FETCH_PROJECT_SUCCESS,
  FETCH_PROJECT_FAILED,
} from "./utils/ActionTypes";
import { success, start, failed } from "./utils/actions";

export function fetchProject(id) {
  return (dispatch) => {
    dispatch(start(FETCH_PROJECT_START));
    axios
      .get(ENDPOINT_PROJECTS + id + "/")
      .then(function (response) {
        dispatch(success(FETCH_PROJECT_SUCCESS, response.data));
      })
      .catch(function (error) {
        dispatch(failed(FETCH_PROJECT_FAILED, error));
      });
  };
}
