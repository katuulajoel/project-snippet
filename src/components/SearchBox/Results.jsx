/* -------------------------------------------------------------------------- */
/*                            External Dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import styled from "styled-components";

/* -------------------------- Internel Dependencies ------------------------- */
import Progress from "../Progress";
import LoadMore from "../LoadMore";
import { listMoreInvoices } from "../../redux/actions/InvoiceActions";
import SummaryPlaceholder from "../SummaryPlaceholder/SummaryPlaceholder";

const Results = ({ searchTerm, clearSearch, navHieght }) => {
  const { search, isMakingRequest } = useSelector(({ Invoice }) => Invoice);
  const dispatch = useDispatch();
  return (
    <>
      {searchTerm && (
        <Wrapper navHieght={navHieght?.current?.offsetHeight}>
          {isMakingRequest.search && <Progress />}
          {search.data.length == 0 ? (
            <SummaryPlaceholder description="No results found" />
          ) : (
            <div className="results">
              {search.data.length && (
                <span className="counter" data-testid="counter">
                  Showing <strong>{search.data.length}</strong> results for{" "}
                  <strong>&quot;{searchTerm}&quot;</strong>
                </span>
              )}

              {isMakingRequest.search ? null : search.data.length ? (
                <div className="section">
                  <div className="title">Payments</div>
                  {search.data.map((invoice) => {
                    return (
                      <Link
                        key={invoice?.id}
                        to={`/projects/${invoice?.project.id}/pay`}
                        className="result-item"
                        onClick={clearSearch}
                      >
                        <b>{invoice?.project.title}</b>: {invoice?.title}
                      </Link>
                    );
                  })}

                  <LoadMore
                    type="text"
                    hasMore={search.next && search.next != ""}
                    isLoadingMore={isMakingRequest.search}
                    onLoadMore={() =>
                      dispatch(listMoreInvoices(search.next, true))
                    }
                    variant="outline-primary"
                  >
                    View More
                  </LoadMore>
                </div>
              ) : null}
            </div>
          )}
        </Wrapper>
      )}
    </>
  );
};

const Wrapper = styled.div`
  width: 413px;
  top: ${(props) => props.navHieght}px;
  height: calc(100vh - ${(props) => props.navHieght + "px"});
  right: 0;
  overflow-y: scroll;

  ${(props) => props?.theme?.mixins?.scrollbar}
  padding: 40px;
  position: absolute;
  background: #fff;
  border: 1px solid #efefef;
  margin-top: 7.0625rem;

  .results {
    font-size: 16px;
    line-height: 19px;
    color: #151a30;

    .counter {
      b {
        font-weight: 600;
      }
    }
    .section {
      padding: 40px 0;
      border-bottom: 1px solid #edf1f7;

      &:last-of-type {
        border-bottom: none;
      }

      .title {
        font-weight: 600;
        margin-bottom: 16px;
      }
      a {
        display: block;
        text-decoration: none;
        font-weight: 500;
        margin-bottom: 16px;
        color: #151a30;
        b {
          font-weight: 600;
        }
      }

      .text-link {
        button {
          border: none;
          padding: 0;
          font-weight: 600;
          font-size: 14px;
          line-height: 17px;
          color: #da3451;
        }
        button:hover,
        button:focus {
          background-color: #fff;
          color: #da3451;
        }
      }

      .avatar.avatar-dash {
        width: 48px;
        height: 48px;

        &.avatar-initials {
          font-size: 16px;
        }
      }
    }
  }
`;

Results.propTypes = {
  searchTerm: PropTypes.string,
  clearSearch: PropTypes.func,
  navHieght: PropTypes.any,
};

export default Results;
