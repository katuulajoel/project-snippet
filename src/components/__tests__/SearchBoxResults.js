import React from "react";
import renderer from "react-test-renderer";
import Results from "../SearchBox/Results";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { mount } from "enzyme/build";
import * as actions from "../../actions/InvoiceActions";

const middlewares = [thunk];

const mockAppState = {
  Invoice: {
    search: {
      data: [{ id: 123, project: { id: 1 } }],
      next: "https://example.com",
    },
    isMakingRequest: {},
  },
};

const mockAppStore = (state) => {
  const mockStore = configureStore(middlewares);
  return mockStore(state || mockAppState);
};

describe("SearchBox Results test", () => {
  it("Snapshot test for search results component", () => {
    const tree = renderer
      .create(
        <Provider store={mockAppStore()}>
          <Router>
            <Results searchTerm="Hello" />
          </Router>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should show list of results from search", () => {
    const listMoreInvoicesStub = jest.spyOn(actions, "listMoreInvoices");
    const wrapper = mount(
      <Provider store={mockAppStore()}>
        <Router>
          <Results searchTerm="Hello" />
        </Router>
      </Provider>
    );

    wrapper.find("LoadMore button").simulate("click");
    expect(listMoreInvoicesStub).toHaveBeenCalledWith(
      "https://example.com",
      true
    );
  });
});
