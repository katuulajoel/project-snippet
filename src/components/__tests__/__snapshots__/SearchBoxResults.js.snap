// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox Results test Snapshot test for search results component 1`] = `
<div
  className="sc-bdnxRM igYzrV"
>
  <div
    className="results"
  >
    <span
      className="counter"
      data-testid="counter"
    >
      Showing 
      <strong>
        1
      </strong>
       results for
       
      <strong>
        "
        Hello
        "
      </strong>
    </span>
    <div
      className="section"
    >
      <div
        className="title"
      >
        Payments
      </div>
      <a
        className="result-item"
        href="/projects/1/pay"
        onClick={[Function]}
      >
        <b />
        : 
      </a>
      <div
        className="text-link"
      >
        <button
          className="btn btn-outline-primary   btn-sm"
          form={null}
          onClick={[Function]}
          type="button"
        >
          View More
        </button>
      </div>
    </div>
  </div>
</div>
`;
